build error with firebird-2.5.7 on nixos 20.09

**Bug description**

nixos-build switch produces a build error due to  firebird-2.5.7.27050-0 while firebird is not explicitly mentioned anywhere in the nixox/congiuration.nix file...

**To Reproduce**
1. nix-channel --update
2. nixos-build switch 
3. build  error message
4. removed a few apps from congiuration.nix file hoping that one of them would be asking for firebird to be installed behind the scene at no avail...

**Expected behavior**
A clean error free build.

**Screenshot**

...
making all in engines/ccgost...
g++ -D_REENTRANT -I. -I../common -I../common -I./unicode -I./.. -D_REENTRANT -DHAVE_CONFIG_H -DU_LAYOUT_IMPLEMENTATION -O  -c -DPIC -fPIC -o HebrewLigatureData.o HebrewLigatureData.cpp
make[2]: Entering directory '/build/openssl-1.0.2u/engines/ccgost'
gcc -I../../include -fPIC -DOPENSSL_PIC -DOPENSSL_THREADS -D_REENTRANT -DDSO_DLFCN -DHAVE_DLFCN_H -m64 -DL_ENDIAN -O3 -Wall -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DRC4_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM   -c -o e_gost_err.o e_gost_err.c
g++ -D_REENTRANT -I. -I../common -I../common -I./unicode -I./.. -D_REENTRANT -DHAVE_CONFIG_H -DU_LAYOUT_IMPLEMENTATION -O  -c -DPIC -fPIC -o HebrewShaping.o HebrewShaping.cpp
g++ -D_REENTRANT -I. -I../common -I../common -I./unicode -I./.. -D_REENTRANT -DHAVE_CONFIG_H -DU_LAYOUT_IMPLEMENTATION -O  -c -DPIC -fPIC -o IndicClassTables.o IndicClassTables.cpp
gcc -I../../include -fPIC -DOPENSSL_PIC -DOPENSSL_THREADS -D_REENTRANT -DDSO_DLFCN -DHAVE_DLFCN_H -m64 -DL_ENDIAN -O3 -Wall -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DRC4_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM   -c -o gost2001_keyx.o gost2001_keyx.c
IndicClassTables.cpp:229:117: error: narrowing conversion of '2147483651' from 'unsigned int' to 'icu_3_0::IndicClassTable::ScriptFlags' {aka 'int'} [-Wnarrowing]
  229 | static const IndicClassTable teluClassTable = {0x0C00, 0x0C6F, 3, TELU_SCRIPT_FLAGS, teluCharClasses, teluSplitTable};
      |                                                                                                                     ^
IndicClassTables.cpp:231:117: error: narrowing conversion of '2147483651' from 'unsigned int' to 'icu_3_0::IndicClassTable::ScriptFlags' {aka 'int'} [-Wnarrowing]
  231 | static const IndicClassTable kndaClassTable = {0x0C80, 0x0CEF, 4, KNDA_SCRIPT_FLAGS, kndaCharClasses, kndaSplitTable};
      |                                                                                                                     ^
make[4]: *** [../config/mh-linux:43: IndicClassTables.o] Error 1
make[4]: Leaving directory '/build/Firebird-2.5.7.27050-0/extern/icu/source/layout'
make[3]: *** [Makefile:98: all-recursive] Error 2
make[3]: Leaving directory '/build/Firebird-2.5.7.27050-0/extern/icu/source'
make[2]: *** [../gen/Makefile.extern.icu:39: libicu] Error 2
make[2]: Leaving directory '/build/Firebird-2.5.7.27050-0/gen'
make[1]: *** [Makefile:193: external_libraries] Error 2
make[1]: Leaving directory '/build/Firebird-2.5.7.27050-0/gen'
make: *** [Makefile:6: firebird] Error 2
builder for '/nix/store/589jgyzjbbn6zl870qswm80pi4xh8ffq-firebird-2.5.7.27050-0.drv' failed with exit code 2
building '/nix/store/f5v53hni9pnldx4l0v4svxv0ibl62ir3-tor-browser-linux64-9.0.9_en-US.tar.xz.drv'...
cannot build derivation '/nix/store/c1wpclz112ssp5ki6m546llszn144bq8-wt-4.2.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/8izfzp2k51zz3385cq14ffd31l2c8bwg-fileshelter-4.0.1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/wi7fzbgmwj9rrzm48mlcsvw51lymmcz8-system-path.drv': 1 dependencies couldn't be built
building '/nix/store/75r17sql3df7jq3azd76w9fjw6870wm6-torrc.drv'...
building '/nix/store/0zq0cb92k0w72ri9jb2zd65gzpnhx6fv-udev-rules.drv'...
building '/nix/store/jqc775s8a0m6r6qhwcfssjdya326s6z0-unit-NetworkManager-dispatcher.service.drv'...
building '/nix/store/bvg0a7ah4bd77s86f16rq7jsl7sh3nxy-unit-NetworkManager-wait-online.service.drv'...
building '/nix/store/vbnnzk9kdx828c8vi3pqcjayrf1ja43n-unit-NetworkManager.service.drv'...
building '/nix/store/df1ffd22pbca5x7ya5mwa8prm40mg54p-unit-alsa-store.service.drv'...
cannot build derivation '/nix/store/4375h9w8pmppld22s4jb6ag05riw87vg-xsession-wrapper.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/gi689m2ph1avv6n2svvlzr5d2v2xm4p1-nixos-system-bx26-20.09pre226012.b47873026c7.drv': 1 dependencies couldn't be built
error: build of '/nix/store/gi689m2ph1avv6n2svvlzr5d2v2xm4p1-nixos-system-bx26-20.09pre226012.b47873026c7.drv' failed




**Notify  @maintainers**
@Marc Weber <marco-oweber@gmx.de>
<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.28, NixOS, 20.09pre221706.b61999e4ad6 (Nightingale)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.4`
 - channels(root): `"nixos-20.09pre226012.b47873026c7"`
 - channels(k): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

--- a/pkgs/development/libraries/libtomcrypt/default.nix
+++ b/pkgs/development/libraries/libtomcrypt/default.nix
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchurl, fetchpatch, libtool }:
+{ lib, stdenv, fetchurl, fetchpatch, libtool, libtommath }:
 
 stdenv.mkDerivation rec {
   pname = "libtomcrypt";
@@ -17,14 +17,16 @@ stdenv.mkDerivation rec {
     })
   ];
 
-  nativeBuildInputs = [ libtool ];
+  nativeBuildInputs = [ libtool libtommath ];
 
   postPatch = ''
     substituteInPlace makefile.shared --replace "LT:=glibtool" "LT:=libtool"
   '';
 
   preBuild = ''
-    makeFlagsArray=(PREFIX=$out \
+    makeFlagsArray+=(PREFIX=$out \
+      CFLAGS="-DUSE_LTM -DLTM_DESC -DLTC_PTHREAD" \
+      EXTRALIBS=\"-ltommath\" \
       INSTALL_GROUP=$(id -g) \
       INSTALL_USER=$(id -u))
   '';

--- a/pkgs/servers/firebird/default.nix
+++ b/pkgs/servers/firebird/default.nix
@@ -1,91 +1,84 @@
-{lib, stdenv, fetchurl, libedit, automake, autoconf, libtool
-,
-  # icu = null: use icu which comes with firebird
+{ lib, stdenv, fetchFromGitHub, libedit, autoreconfHook, zlib, unzip, libtommath, libtomcrypt, icu, superServer ? false }:
 
-  # icu = pkgs.icu => you may have trouble sharing database files with windows
-  # users if "Collation unicode" columns are being used
-  # windows icu version is *30.dll, however neither the icu 3.0 nor the 3.6
-  # sources look close to what ships with this package.
-  # Thus I think its best to trust firebird devs and use their version
-
-  # icu version missmatch may cause such error when selecting from a table:
-  # "Collation unicode for character set utf8 is not installed"
-
-  # icu 3.0 can still be built easily by nix (by dropping the #elif case and
-  # make | make)
-  icu ? null
-
-, superServer ? false
-, port ? 3050
-, serviceName ? "gds_db"
-}:
-
-/*
-   there are 3 ways to use firebird:
-   a) superserver
-    - one process, one thread for each connection
-   b) classic
-    - is built by default
-    - one process for each connection
-    - on linux direct io operations (?)
-   c) embedded.
-
-   manual says that you usually don't notice the difference between a and b.
-
-   I'm only interested in the embedder shared libary for now.
-   So everything isn't tested yet
-
-*/
-
-stdenv.mkDerivation rec {
-  version = "2.5.7.27050-0";
+let base = {
   pname = "firebird";
 
-  # enableParallelBuilding = false; build fails
+  meta = with lib; {
+    description = "SQL relational database management system";
+    homepage = "https://firebirdsql.org/";
+    changelog = "https://github.com/FirebirdSQL/firebird/blob/master/CHANGELOG.md";
+    license = [ "IDPL" "Interbase-1.0" ];
+    platforms = platforms.linux;
+    maintainers = with maintainers; [ marcweber ];
+  };
 
-  # http://tracker.firebirdsql.org/browse/CORE-3246
-  preConfigure = ''
-    makeFlags="$makeFlags CPU=$NIX_BUILD_CORES"
+  nativeBuildInputs = [ autoreconfHook ];
+
+  buildInputs = [ libedit icu ];
+
+  LD_LIBRARY_PATH = lib.makeLibraryPath [ icu ];
+
+  configureFlags = [
+    "--with-system-editline"
+  ] ++ (lib.optional superServer "--enable-superserver");
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out
+    cp -r gen/Release/firebird/* $out
+    runHook postInstall
   '';
 
-  configureFlags =
-    [ "--with-serivec-port=${builtins.toString port}"
-      "--with-service-name=${serviceName}"
-      "--with-system-editline"
-      "--with-fblog=/var/log/firebird"
-      "--with-fbconf=/etc/firebird"
-      "--with-fbsecure-db=/var/db/firebird/system"
-    ]
-    ++ (lib.optional  (icu != null) "--with-system-icu")
-    ++ (lib.optional superServer "--enable-superserver");
+}; in {
 
-  src = fetchurl {
-    url = "mirror://sourceforge/firebird/Firebird-${version}.tar.bz2";
-    sha256 = "06hp6bq5irqvm3h03s79qjgcc3jsjpq150y9aq7anklx9v4nhfqa";
-  };
+  firebird_2_5 = stdenv.mkDerivation (base // rec {
+    version = "2.5.9";
 
-  hardeningDisable = [ "format" ];
+    src = fetchFromGitHub {
+      owner = "FirebirdSQL";
+      repo = "firebird";
+      rev = "R${builtins.replaceStrings [ "." ] [ "_" ] version}";
+      sha256 = "sha256-YyvlMeBux80OpVhsCv+6IVxKXFRsgdr+1siupMR13JM=";
+    };
 
-  # configurePhase = ''
-  #   sed -i 's@cp /usr/share/automake-.*@@' autogen.sh
-  #   sh autogen.sh $configureFlags --prefix=$out
-  # '';
-  buildInputs = [libedit icu automake autoconf libtool];
+    configureFlags = base.configureFlags ++ [ "--with-system-icu" ];
 
-  # TODO: Probably this hase to be tidied up..
-  # make install requires beeing. disabling the root checks
-  # dosen't work. Copying the files manually which can be found
-  # in ubuntu -dev -classic, -example packages:
-  # maybe some of those files can be removed again
-  installPhase = "cp -r gen/firebird $out";
+    installPhase = ''
+      runHook preInstall
+      mkdir -p $out
+      cp -r gen/firebird/* $out
+      runHook postInstall
+    '';
 
-  meta = {
-    description = "SQL relational database management system";
-    homepage = "https://www.firebirdnews.org";
-    license = ["IDPL" "Interbase-1.0"];
-    maintainers = [lib.maintainers.marcweber];
-    platforms = lib.platforms.linux;
-    broken = true;
-  };
+    meta = base.meta // { platforms = [ "x86_64-linux" ]; };
+  });
+
+  firebird_3 = stdenv.mkDerivation (base // rec {
+    version = "3.0.7";
+
+    src = fetchFromGitHub {
+      owner = "FirebirdSQL";
+      repo = "firebird";
+      rev = "R${builtins.replaceStrings [ "." ] [ "_" ] version}";
+      sha256 = "sha256-8nGan10qjW8dFF89BL/tUWtwMGhahBiODbOqRrHSrbs=";
+    };
+
+    buildInputs = base.buildInputs ++ [ zlib libtommath ];
+
+    meta = base.meta // { platforms = [ "x86_64-linux" ]; };
+  });
+
+  firebird_4 = stdenv.mkDerivation (base // rec {
+    version = "4.0.0-rc1";
+
+    src = fetchFromGitHub {
+      owner = "FirebirdSQL";
+      repo = "firebird";
+      rev = "T4_0_0_RC1";
+      sha256 = "sha256-FLaRePosF5dtJ+fmrfvzkE6wawC9Z9YLhT/ZWkwWXb4=";
+    };
+
+    buildInputs = base.buildInputs ++ [ zlib unzip libtommath libtomcrypt ];
+  });
 
 }

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -18775,8 +18775,8 @@ in
 
   fingerd_bsd = callPackage ../servers/fingerd/bsd-fingerd { };
 
-  firebird = callPackage ../servers/firebird { icu = null; /*stdenv = gcc5Stdenv;*/ };
-  firebirdSuper = firebird.override { icu = icu58; superServer = true; /*stdenv = gcc5Stdenv;*/ };
+  inherit (callPackages ../servers/firebird { }) firebird_4 firebird_3 firebird_2_5;
+  firebird = firebird_3;
 
   freeradius = callPackage ../servers/freeradius { };
 

